# Assignment 3

In this assignment, we will focus on healthcare. This data set is made available by MIT. It contains data about 9,026 heartbeat measurements. Each row represents a single measurement (captured on a timeline). There are a total of 80 data points (columns). This is a multiclass classification task: predict whether the measurement represents a normal heartbeat or other anomalies. 

## Description of Variables

You will use the **hearbeat_cleaned.csv** data set for this assignment. Each row represents a single measurement. Columns labeled as T1 from T80 are the time steps on the timeline (there are 80 time steps, each time step has only one measurement). 

The last column is the target variable. It shows the label (category) of the measurement as follows:<br>
0 = Normal<br>
1 = Supraventricular premature beat<br>
2 = Premature ventricular contraction<br>
3 = Fusion of ventricular and normal beat<br>
4 = Unclassifiable beat

## Goal

Use the data set **hearbeat_cleaned.csv** to predict the column called **Target**. The input variables are columns labeled as **T1 to T80**. 

## Submission:

Please save and submit this Jupyter notebook file. The correctness of the code matters for your grade. **Readability and organization of your code is also important.** You may lose points for submitting unreadable/undecipherable code. Therefore, use markdown cells to create sections, and use comments where necessary.


# Note:

The data is cleaned up. There are no unqueal length sequences. And, there is no zero padding. So, you shouldn't use any `Masking` layer (like I mentioned in the lecture). 

# Read and Prepare the Data (1 points)

In [ ]:
# Insert as many cells as you need for data prep

In [2]:
#Common imports
import numpy as np
import pandas as pd
from tensorflow import keras
import tensorflow as tf


## Get the data

In [3]:
data = pd.read_csv("heartbeat_cleaned.csv")


In [4]:
data.shape

(7960, 81)

In [5]:
data.head()

T1     T2     T3      T4      T5      T6      T7      T8      T9  \
0  0.987  0.892  0.461  0.1130  0.1490  0.1900  0.1650  0.1620  0.1470   
1  1.000  0.918  0.621  0.1330  0.1050  0.1250  0.1170  0.0898  0.0703   
2  1.000  0.751  0.143  0.1040  0.0961  0.0519  0.0442  0.0416  0.0364   
3  1.000  0.740  0.235  0.0464  0.0722  0.0567  0.0103  0.0155  0.0284   
4  1.000  0.833  0.309  0.0191  0.1010  0.1200  0.1040  0.0874  0.0765   

      T10  ...     T72     T73     T74     T75    T76     T77     T78    T79  \
0  0.1380  ...  0.1970  0.1970  0.1960  0.2030  0.201  0.1990  0.2010  0.205   
1  0.0781  ...  0.1950  0.1910  0.1520  0.1720  0.207  0.2110  0.2070  0.207   
2  0.0857  ...  0.2260  0.2420  0.2440  0.2860  0.468  0.8160  0.9770  0.452   
3  0.0155  ...  0.0851  0.0747  0.0515  0.0593  0.067  0.0361  0.1210  0.451   
4  0.0765  ...  0.2050  0.4210  0.8030  0.9510  0.467  0.0000  0.0519  0.082   

      T80  Target  
0  0.2080       0  
1  0.1720       0  
2  0.0519       0  
3  0.8690       0  
4  0.0628       0  

[5 rows x 81 columns]

In [6]:
y = data['Target']
x = data.drop('Target', axis=1)

## Split data

In [7]:
from sklearn.model_selection import train_test_split

train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3)

## Data Transformation

In [8]:
#Target variables need to be an array with integer type
train_y = np.array(train_y)
test_y = np.array(test_y)

train_y = train_y.astype(np.int32)
test_y = test_y.astype(np.int32)


In [9]:
#Check the first 10 values of the train_y data set
train_y[0:10]

array([0, 0, 0, 0, 0, 4, 0, 0, 0, 4])

In [10]:
train_x=np.array(train_x)
test_x=np.array(test_x)
train_x=train_x.astype(np.float32)
test_x=test_x.astype(np.float32)

In [11]:
train_x

array([[0.856 , 0.824 , 0.562 , ..., 0.227 , 0.243 , 0.23  ],
       [1.    , 0.64  , 0.053 , ..., 0.966 , 0.735 , 0.106 ],
       [1.    , 0.754 , 0.0673, ..., 0.219 , 0.226 , 0.229 ],
       ...,
       [0.507 , 0.311 , 0.    , ..., 0.129 , 0.119 , 0.133 ],
       [0.942 , 0.918 , 0.74  , ..., 0.113 , 0.0988, 0.0819],
       [0.872 , 0.864 , 0.791 , ..., 0.279 , 0.294 , 0.286 ]],
      dtype=float32)

In [12]:
train_x = np.reshape(train_x, (train_x.shape[0], train_x.shape[1], 1))
test_x = np.reshape(test_x, (test_x.shape[0], test_x.shape[1], 1))

In [13]:
train_x.shape, train_y.shape

((5572, 80, 1), (5572,))

In [14]:

train_x

array([[[0.856 ],
        [0.824 ],
        [0.562 ],
        ...,
        [0.227 ],
        [0.243 ],
        [0.23  ]],

       [[1.    ],
        [0.64  ],
        [0.053 ],
        ...,
        [0.966 ],
        [0.735 ],
        [0.106 ]],

       [[1.    ],
        [0.754 ],
        [0.0673],
        ...,
        [0.219 ],
        [0.226 ],
        [0.229 ]],

       ...,

       [[0.507 ],
        [0.311 ],
        [0.    ],
        ...,
        [0.129 ],
        [0.119 ],
        [0.133 ]],

       [[0.942 ],
        [0.918 ],
        [0.74  ],
        ...,
        [0.113 ],
        [0.0988],
        [0.0819]],

       [[0.872 ],
        [0.864 ],
        [0.791 ],
        ...,
        [0.279 ],
        [0.294 ],
        [0.286 ]]], dtype=float32)

# Find the baseline (0.5 point)

In [15]:
data['Target'].value_counts()/len(data)

0    0.582035
4    0.198995
2    0.155402
1    0.055905
3    0.007663
Name: Target, dtype: float64

# Build a cross-sectional shallow model using Keras (with only one hidden layer) (2 points)

In [43]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[80, 1]),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

In [44]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y))

Epoch 1/50
175/175 [==============================] - 1s 3ms/step - loss: 0.6992 - accuracy: 0.7687 - val_loss: 0.5864 - val_accuracy: 0.8212
Epoch 2/50
175/175 [==============================] - 0s 2ms/step - loss: 0.4587 - accuracy: 0.8494 - val_loss: 0.4395 - val_accuracy: 0.8710
Epoch 3/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3896 - accuracy: 0.8787 - val_loss: 0.3874 - val_accuracy: 0.8765
Epoch 4/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3551 - accuracy: 0.8945 - val_loss: 0.3100 - val_accuracy: 0.9104
Epoch 5/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3301 - accuracy: 0.9009 - val_loss: 0.3212 - val_accuracy: 0.9133
Epoch 6/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3071 - accuracy: 0.9072 - val_loss: 0.2958 - val_accuracy: 0.9183
Epoch 7/50
175/175 [==============================] - 0s 2ms/step - loss: 0.2968 - accuracy: 0.9083 - val_loss: 0.2839 - val_accuracy: 0.9204
Epoch 

In [45]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.26684197783470154, 0.9292294979095459]

In [46]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.27
accuracy: 92.92%


# Build a cross-sectional deep model using Keras (with two or more hidden layers) (2 points)

In [47]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[80, 1]),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

In [48]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y))

Epoch 1/50
175/175 [==============================] - 1s 3ms/step - loss: 0.6839 - accuracy: 0.7599 - val_loss: 0.4867 - val_accuracy: 0.8183
Epoch 2/50
175/175 [==============================] - 0s 2ms/step - loss: 0.4424 - accuracy: 0.8501 - val_loss: 0.4454 - val_accuracy: 0.8463
Epoch 3/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3694 - accuracy: 0.8790 - val_loss: 0.4002 - val_accuracy: 0.8702
Epoch 4/50
175/175 [==============================] - 0s 2ms/step - loss: 0.3243 - accuracy: 0.8979 - val_loss: 0.2659 - val_accuracy: 0.9171
Epoch 5/50
175/175 [==============================] - 0s 2ms/step - loss: 0.2996 - accuracy: 0.9020 - val_loss: 0.2788 - val_accuracy: 0.9221
Epoch 6/50
175/175 [==============================] - 0s 2ms/step - loss: 0.2796 - accuracy: 0.9078 - val_loss: 0.3169 - val_accuracy: 0.8874
Epoch 7/50
175/175 [==============================] - 0s 2ms/step - loss: 0.2534 - accuracy: 0.9191 - val_loss: 0.2710 - val_accuracy: 0.9229
Epoch 

In [40]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.2824445962905884, 0.908710241317749]

In [36]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.28
accuracy: 90.87%


# Build a sequential shallow LSTM Model (with only one LSTM layer) (2 points)

In [23]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    
    keras.layers.LSTM(20, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='softmax')
])

In [24]:
from tensorflow.keras.callbacks import EarlyStopping


earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [25]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
175/175 [==============================] - 6s 24ms/step - loss: 1.1176 - accuracy: 0.5765 - val_loss: 1.1715 - val_accuracy: 0.5900
Epoch 2/20
175/175 [==============================] - 4s 23ms/step - loss: 1.0660 - accuracy: 0.5876 - val_loss: 1.0731 - val_accuracy: 0.5892
Epoch 3/20
175/175 [==============================] - 4s 23ms/step - loss: 1.0547 - accuracy: 0.5953 - val_loss: 0.9616 - val_accuracy: 0.6131
Epoch 4/20
175/175 [==============================] - 4s 22ms/step - loss: 0.9178 - accuracy: 0.6554 - val_loss: 1.4403 - val_accuracy: 0.4824
Epoch 5/20
175/175 [==============================] - 4s 22ms/step - loss: 0.7674 - accuracy: 0.7349 - val_loss: 1.4447 - val_accuracy: 0.4376
Epoch 6/20
175/175 [==============================] - 4s 23ms/step - loss: 0.7908 - accuracy: 0.7340 - val_loss: 2.0000 - val_accuracy: 0.2818
Epoch 7/20
175/175 [==============================] - 4s 22ms/step - loss: 0.7038 - accuracy: 0.7624 - val_loss: 0.5684 - val_accuracy: 0.8166

In [49]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.5477679967880249, 0.8907034993171692]

In [50]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.55
accuracy: 89.07%


# Build a sequential deep LSTM Model (with only two LSTM layers) (2 points)

In [51]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.LSTM(20, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.LSTM(20),
    keras.layers.Dense(5, activation='softmax')
])

In [52]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
175/175 [==============================] - 12s 53ms/step - loss: 1.1276 - accuracy: 0.5799 - val_loss: 1.0602 - val_accuracy: 0.6068
Epoch 2/20
175/175 [==============================] - 10s 57ms/step - loss: 1.0990 - accuracy: 0.5833 - val_loss: 1.0898 - val_accuracy: 0.5900
Epoch 3/20
175/175 [==============================] - 9s 50ms/step - loss: 1.0174 - accuracy: 0.6071 - val_loss: 1.0847 - val_accuracy: 0.6106
Epoch 4/20
175/175 [==============================] - 9s 49ms/step - loss: 1.0106 - accuracy: 0.6019 - val_loss: 1.0166 - val_accuracy: 0.5900
Epoch 5/20
175/175 [==============================] - 10s 59ms/step - loss: 1.0768 - accuracy: 0.5897 - val_loss: 1.0187 - val_accuracy: 0.6013
Epoch 6/20
175/175 [==============================] - 9s 52ms/step - loss: 0.9993 - accuracy: 0.6145 - val_loss: 0.9677 - val_accuracy: 0.6382
Epoch 7/20
175/175 [==============================] - 9s 51ms/step - loss: 0.9955 - accuracy: 0.6416 - val_loss: 0.9502 - val_accuracy: 0.6

In [53]:
scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.5145398378372192, 0.8094639778137207]

In [54]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.51
accuracy: 80.95%


# Build a sequential shallow GRU Model (with only one GRU layer) (2 points)

In [55]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(20, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='Softmax')
])

In [56]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
175/175 [==============================] - 7s 30ms/step - loss: 1.1066 - accuracy: 0.5838 - val_loss: 1.1285 - val_accuracy: 0.5771
Epoch 2/20
175/175 [==============================] - 6s 36ms/step - loss: 0.9006 - accuracy: 0.6744 - val_loss: 0.6971 - val_accuracy: 0.7596
Epoch 3/20
175/175 [==============================] - 5s 29ms/step - loss: 0.6233 - accuracy: 0.7929 - val_loss: 0.6020 - val_accuracy: 0.7864
Epoch 4/20
175/175 [==============================] - 7s 40ms/step - loss: 0.4948 - accuracy: 0.8424 - val_loss: 0.4995 - val_accuracy: 0.8132
Epoch 5/20
175/175 [==============================] - 6s 34ms/step - loss: 0.4266 - accuracy: 0.8632 - val_loss: 0.3933 - val_accuracy: 0.8790
Epoch 6/20
175/175 [==============================] - 5s 30ms/step - loss: 0.4196 - accuracy: 0.8688 - val_loss: 0.4050 - val_accuracy: 0.8865
Epoch 7/20
175/175 [==============================] - 6s 34ms/step - loss: 0.3892 - accuracy: 0.8803 - val_loss: 0.3498 - val_accuracy: 0.8995

In [57]:
scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.3676387667655945, 0.8823283314704895]

In [58]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.37
accuracy: 88.23%


# Build a sequential deep GRU Model (with only two GRU layers) (2 points)

In [62]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(20, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.GRU(20),
    keras.layers.Dense(5, activation='Softmax')
])

In [63]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=20,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/20
175/175 [==============================] - 7s 28ms/step - loss: 1.0858 - accuracy: 0.5903 - val_loss: 1.0594 - val_accuracy: 0.5842
Epoch 2/20
175/175 [==============================] - 5s 26ms/step - loss: 0.8601 - accuracy: 0.6814 - val_loss: 0.7184 - val_accuracy: 0.7316
Epoch 3/20
175/175 [==============================] - 5s 26ms/step - loss: 0.5442 - accuracy: 0.8232 - val_loss: 0.7250 - val_accuracy: 0.7295
Epoch 4/20
175/175 [==============================] - 5s 26ms/step - loss: 0.4842 - accuracy: 0.8464 - val_loss: 0.3710 - val_accuracy: 0.8907
Epoch 5/20
175/175 [==============================] - 4s 26ms/step - loss: 0.4029 - accuracy: 0.8746 - val_loss: 0.2989 - val_accuracy: 0.9125
Epoch 6/20
175/175 [==============================] - 5s 26ms/step - loss: 0.3793 - accuracy: 0.8837 - val_loss: 0.3247 - val_accuracy: 0.9008
Epoch 7/20
175/175 [==============================] - 4s 26ms/step - loss: 0.3191 - accuracy: 0.9011 - val_loss: 0.2941 - val_accuracy: 0.9091

In [64]:
scores = model.evaluate(test_x, test_y, verbose=0)

scores

[0.2824445962905884, 0.908710241317749]

In [65]:
print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.28
accuracy: 90.87%


# Discussion

## List the test values of each model you built (0.5 points)

# Which model performs the best and why? (0.5 points) 
## How does it compare to baseline? (0.5 points)



# 

Cross sectional Shallow Keras Model performs the best when compared to the remaining models. Its Test Accuracy is 92.2% which remains highest than all other test accuracies.

Reason: 1.Number of neurons & hidden layers in shallow Keras Model with appropriate learning rate and Epochs helps the model to get the best test accuracy. 

Comparison with Baseline: 0.5825

Cross sectional Shallow Keras Model has 92.2% test accuracy which is way larger value than 58.25%. Evidently we can conclude that it is a efficient Model.